benchmark/rust-ffi/app/Main.hs

```diff
@@ -0,0 +1,95 @@
+module Main where
+
+import Control.DeepSeq (force)
+import Control.Exception (evaluate)
+import Control.Monad (void, when)
+import Data.Foldable (for_, traverse_)
+import Data.Word (Word8)
+import Foreign.C.Types (CBool (..), CSize (..))
+import Foreign.Marshal (with, withArray)
+import Foreign.Ptr (Ptr, castPtr)
+import Foreign.Storable (Storable (sizeOf), alignment, peek, poke, pokeByteOff, sizeOf)
+import GHC.Compact (compact, getCompact)
+import System.IO (hPrint, stderr, stdout)
+import Text.Printf (hPrintf)
+
+import qualified Data.Primitive.ByteArray as BA
+import qualified System.Clock as Clock
+import qualified System.Environment as Env
+
+import Data.Text.Utf8 (CodeUnitIndex (..), Text (..))
+
+import qualified Data.Text.Utf8 as Utf8
+
+foreign import ccall unsafe "perform_ac"
+  performAc :: CBool -> CSize -> Ptr U8Slice -> Ptr U8Slice -> IO CSize
+
+data U8Slice = U8Slice (Ptr Word8) CSize CSize
+
+instance Storable U8Slice where
+  sizeOf _ = sizeOf (undefined :: Ptr Word8) + 2 * sizeOf (undefined :: CSize)
+  alignment _ = max (alignment (undefined :: Ptr Word8)) (alignment (undefined :: CSize))
+  peek _ptr = error "We only write U8Slices to pointer, never read them"
+  poke ptr (U8Slice u8ptr off len) = do
+    poke ptr' u8ptr
+    pokeByteOff ptr' (sizeOf (undefined :: Ptr Word8)) off
+    pokeByteOff ptr' (sizeOf (undefined :: Ptr Word8) + sizeOf (undefined :: CSize)) len
+    where ptr' = castPtr ptr
+
+-- | Turn a 'Text' value into something that can be passed through FFI.
+-- The 'ByteArray' values inside must be pinned.
+fromText :: Text -> U8Slice
+fromText (Text u8data off len)
+  | BA.isByteArrayPinned u8data = U8Slice (BA.byteArrayContents u8data) (fromIntegral off) (fromIntegral len)
+  | otherwise                   = error "Won't pass an unpinned ByteArray through FFI"
```

Comment:
Small technicality:

> Won't pass an unpinned ByteArray through FFI

Passing an unpinned array through FFI (when using an unsafe FFI call) is not a problem (see https://downloads.haskell.org/~ghc/8.10.4/docs/html/users_guide/ffi-chap.html#unlifted-ffi-types). The issue here is that we're *not* (yet) passing it through FFI at this point, we take the address but stay in Haskell-land.